backpack/core/derivatives/batchnorm_nd.py

```diff
@@ -0,0 +1,287 @@
+        if module.input0.dim() != 2:
+            raise NotImplementedError(
+                "residual_mat_prod is implemented only for 0 dimensions. "
+                "If you need more dimension make a feature request."
+            )
+
+        N = module.input0.size(0)
+        x_hat, var = self._get_normalized_input_and_var(module)
+        gamma = module.weight
+        eps = module.eps
+
+        factor = gamma / (N * (var + eps))
+
+        sum_127 = einsum("nc,vnc->vc", x_hat, mat)
+        sum_24 = einsum("nc->c", g_out[0])
+        sum_3 = einsum("nc,vnc->vc", g_out[0], mat)
+        sum_46 = einsum("vnc->vc", mat)
+        sum_567 = einsum("nc,nc->c", x_hat, g_out[0])
+
+        r_mat = -einsum("nc,vc->vnc", g_out[0], sum_127)
+        r_mat += (1.0 / N) * einsum("c,vc->vc", sum_24, sum_127).unsqueeze(1).expand(
+            -1, N, -1
+        )
+        r_mat -= einsum("nc,vc->vnc", x_hat, sum_3)
+        r_mat += (1.0 / N) * einsum("nc,c,vc->vnc", x_hat, sum_24, sum_46)
+
+        r_mat -= einsum("vnc,c->vnc", mat, sum_567)
+        r_mat += (1.0 / N) * einsum("c,vc->vc", sum_567, sum_46).unsqueeze(1).expand(
+            -1, N, -1
+        )
+        r_mat += (3.0 / N) * einsum("nc,vc,c->vnc", x_hat, sum_127, sum_567)
+
+        return einsum("c,vnc->vnc", factor, r_mat)
+
+    ###############################################################
+    #                  HELPER FUNCTIONS                         ###
+    ###############################################################
+    def _get_normalized_input_and_var(
+        self, module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d]
+    ) -> Tuple[Tensor, Tensor]:
+        input: Tensor = module.input0
+        if module.training:
+            dim: Tuple[int] = self._get_free_axes(module, index_batch=0)
+            mean: Tensor = input.mean(dim=dim)
+            var: Tensor = input.var(dim=dim, unbiased=False)
+        else:
+            mean: Tensor = module.running_mean
+            var: Tensor = module.running_var
+        mean: Tensor = self._unsqueeze_free_axis(module, mean, 0)
+        var_expanded: Tensor = self._unsqueeze_free_axis(module, var, 0)
+        return (input - mean) / (var_expanded + module.eps).sqrt(), var
+
+    def _get_denominator(
+        self, module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d]
+    ) -> int:
+        shape_input: Size = module.input0.shape
+        free_axes: Tuple[int] = self._get_free_axes(module, index_batch=0)
+        denominator: int = 1
+        for index in free_axes:
+            denominator *= shape_input[index]
+        return denominator
+
+    @staticmethod
+    def _get_n_axis(module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d]) -> int:
+        return module.input0.dim() - 2
+
+    def _unsqueeze_free_axis(
+        self,
+        module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d],
+        tensor: Tensor,
+        index_batch: int,
+    ) -> Tensor:
```

Comment:
Add docstring, mention that this avoids broadcasting.

Reply:
added docstring